MCTP Bridge Polling

faizana-nvidia opened the pull request:
**MCTP Bridge Endpoint poll**

Reference :  DSP0236 section 8.17.6 Reclaiming EIDs from hot-plug devices

```The bus owner/bridge can detect a removed device or devices by validating the EIDs that are presently allocated to endpoints that are directly on the bus and identifying which EIDs are missing. It can do this by attempting to access each endpoint that the bridge has listed in its routing table as being a device that is directly on the particular bus. Attempting to access each endpoint can be accomplished by issuing the Get Endpoint ID command.```

This PR aims to extend ```mctpd``` as bus owner to enhance bridge support feature by incorporating downstream endpoint polling.
One way to detect the accessibility of its downstream devices is for bus owner/bridge to probes on each bridge's downstream endpoint using the Get Endpoint ID command.

Once its established that endpoint path is accessible, we would stop polling to that eid and publish the eid endpoint object

This PR is based  discussions over MCTP Bridge Support : https://github.com/CodeConstruct/mctp/pull/71

faizana-nvidia commented:
I also have added a TODO comment here [todo](https://github.com/CodeConstruct/mctp/pull/85/commits/955d1b4c08fe89b2cc02620820125ce4e0a6b5ce#diff-222eb3ffb54581db6e3d72a6efa39bf25a8145a7f8eab9b7f4474ea184675da9R3790)

in our current implementation of Busowner pool and local eid assignment, we don't have any check to see, if localEID which is being set via mctp tool **(mctp addr add \<eid> dev \<interface>)** for any network, has already been used by Busowner when it assigned dynamic/static EID to any MCTP device/endpoint. This makes current behavior mess up the network peers[] structure as overall Busowner EID space is common to local and endpoint EIDs. I have been meaning to raise an issue to ask this question.

Since we are getting into the reservation of EIDs, our new localEID could very well be from the reserved EID space. So should we introduce restriction of such manner in mctp tool?

jk-ozlabs commented:
> May be we can introduce a new var representing actual_pool_size (response of SET_ENDPOINT ID) and call out these two vars, peer->pool_start and peer->pool_size as allocated data while network reserved_eid_set continues to interact with these vars as it has been doing.

Yes, that's along the lines of what I was thinking. But maybe with the names adjusted: `pool_start` & `pool_size` would represent actually-allocated values, `req_pool_size` and `req_pool_start` for the (temporarily-valid) request details.

And then as an improvement: since the `req_` values are transient, is there any better way to pass them from the Set Endpoint ID response handler to the Allocate Endpoint IDs command? (ie., rather than setting transient values on the struct peer)

> I also have added a TODO comment here 

There's not much we can do about that case; mctpd isn't (currently) managing that local EID, so there is not much we can do on conflicts - as you have noted.

> So should we introduce restriction of such manner in mctp tool/mctpd?

No, not while mctpd is not responsible for assigning local EIDs.

I see the best way to prevent this case is a facility to explicitly configure mctpd with the range of dynamically-allocatable EIDs, and it's up to the platform integrator to ensure that that range does not conflict with the potential set of local EIDs.

(that's one of the reasons for the configuration file support, for exactly this use-case)

faizana-nvidia commented:
> And then as an improvement: since the `req_` values are transient, is there any better way to pass them from the Set Endpoint ID response handler to the Allocate Endpoint IDs command? (ie., rather than setting transient values on the struct peer)

1) To add further onto this, we don't know how many birdges on what bindings would turn up on network. Predicting their numbers would rather be tedious, so would be passing each pool size to their respective Allocate EndpointID. So in worse case scenario assuming 255 bridges per network(similar to peer numbers) we can have a separate array structure representing bridge data/pool_size which would be filled right after SET_ENDPOINT_ID.


2) One another way (though not tested) is somehow find the capacity of bridge just before asking ALLOCATE_ENDPOINT_ID. Unfortunately we need to know this because we want to avoid asking endpoints number in request which would be more that Bridge's dynamic pool size.

Table 23 – Allocate Endpoint IDs message

_An error completion code (ERROR_INVALID_DATA should be returned) shall be returned if the number of EIDs being allocated (Number of Endpoint IDs) exceeds the Dynamic Endpoint ID Pool size. (This error condition does not apply to when the number of endpoint IDs passed in the request is 0x00)._ 


3) We can also think this response of pool_size in SET_EDNPOINT_ID (transient) to be the capacity of bridge while the allocated pool size and start representing allocated values

> > I also have added a TODO comment here
> 
> There's not much we can do about that case; mctpd isn't (currently) managing that local EID, so there is not much we can do on conflicts - as you have noted.
> 
> > So should we introduce restriction of such manner in mctp tool/mctpd?
> 
> No, not while mctpd is not responsible for assigning local EIDs.
> 
> I see the best way to prevent this case is a facility to explicitly configure mctpd with the range of dynamically-allocatable EIDs, and it's up to the platform integrator to ensure that that range does not conflict with the potential set of local EIDs.
> 
> (that's one of the reasons for the configuration file support, for exactly this use-case)

Acknowledged.

jk-ozlabs commented:
> we can have a separate array structure representing bridge data/pool_size which would be filled right after SET_ENDPOINT_ID.

If there is no way to better handle the transient requested peer size, I would prefer to just store it on the `struct peer`.

> One another way (though not tested) is somehow find the capacity of bridge just before asking ALLOCATE_ENDPOINT_ID

We have the capacity of the brige, from the Set Endpoint ID response, no? As you say:

> We can also think this response of pool_size in SET_EDNPOINT_ID (transient) to be the capacity of bridge while the allocated pool size and start representing allocated values

I would suggest:

 1. record the requested pool size from the Set Endpoint ID response
 2. attempt to allocate that requested pool size (pending any configured limits), by querying the `struct peer` array
 3. record the allocated pool range on that peer
 4. perform Allocate Endpoint IDs from that allocated pool

If the allocation is not possible in (2), just fail for now. We can look at fallbacks in a follow-up.

I think this it fine to do without a new dbus call, but am open to ideas about why we may need one.

faizana-nvidia commented:
> > we can have a separate array structure representing bridge data/pool_size which would be filled right after SET_ENDPOINT_ID.
> 
> If there is no way to better handle the transient requested peer size, I would prefer to just store it on the `struct peer`.
> 
> > One another way (though not tested) is somehow find the capacity of bridge just before asking ALLOCATE_ENDPOINT_ID
> 
> We have the capacity of the brige, from the Set Endpoint ID response, no? As you say:
> 
> > We can also think this response of pool_size in SET_EDNPOINT_ID (transient) to be the capacity of bridge while the allocated pool size and start representing allocated values
> 
> I would suggest:
> 
> 1. record the requested pool size from the Set Endpoint ID response
> 2. attempt to allocate that requested pool size (pending any configured limits), by querying the `struct peer` array
> 3. record the allocated pool range on that peer
> 4. perform Allocate Endpoint IDs from that allocated pool
> 
> If the allocation is not possible in (2), just fail for now. We can look at fallbacks in a follow-up.
> 
> I think this it fine to do without a new dbus call, but am open to ideas about why we may need one.

Please let me know if I understood you correctly on steps
1. Fetch out the transient pool size (via SET_ENDPOINT_ID).
2. Try to allocate (find the pool sized chunked of peer structure which are available to be later used by downstream EID via get_pool_start())
3. If found then update the peer structure with req_pool_size = capacity and req_pool_start = get_pool_start()
4. Also mark all such eids to be reserved
5. From dbus api AssignBridgeStatic get pool_size and try to allocate that size with req_pool_start and whatever is coming as response from ALLOCATE_EID command update peer structure with alloc_pool_size = response(pool_size) and alloc_pool_start=response(pool_start).

Introduces req_pool_size and req_pool_start, updates pool_size -> alloc_pool_size and pool_start ->alloc_pool_start.

If we are going by this then do we have to get pool_start argument from dbus method AssignBridgeStatic if we have already decided what pool start will be from req_pool_start?

Or do you mean not to use any new dbus method AssignBridgeStatic at all?
> I think this it fine to do without a new dbus call, but am open to ideas about why we may need one.

jk-ozlabs commented:
Since the allocation mechanism is now in #71, I'll shift this discussion to there.

faizana-nvidia commented:
Hello Jeremy,
I've rebased the PR and updated the logic based on other PR [71](https://github.com/CodeConstruct/mctp/pull/71) comments. Please help me with the review.

jk-ozlabs commented:
I'm going to hold off on reviewing this one until we have the core bridge support done, given that's a hard dependency for this.

faizana-nvidia commented:
> I'm going to hold off on reviewing this one until we have the core bridge support done, given that's a hard dependency for this.

understood, sounds good

faizana-nvidia commented:
Hello Jeremy,
I have some questions on discussed approach of Polling [(discussion)](https://github.com/CodeConstruct/mctp/pull/71#issuecomment-2933806456) mechanism for MCTP Bridge downstream endpoint management.
To give you a better picture on MCTP Network, we have MCU as MCTP Bridge(over USB) managing GPUs at its downstream 
```mermaid
graph LR
    A["HMC/BMC"] --> B["MCTP Bridge"]
    B --> C["GPU1"]
    B --> D["GPU2"]                                          
```
I'm aware that in the [(discussed)](https://github.com/CodeConstruct/mctp/pull/71#issuecomment-2933806456) polling approach, mctpd doesn't care about state of devices behind the bridge once they have been discovered.
In our platform many of our applications such as Nvidia NSM (used for GPU Telemetry), relies directly on Event Subscription model or something similar which creates a requirement to explicitly do a re-subscribe to endpoints again when they undergoing reset. Owing to this and among other similar reasons, this current PR ([MCTP Bridge Polling](https://github.com/CodeConstruct/mctp/pull/85) )proposed for keeping the polling active throughout the existence of MCTP Bridge itself and inform to other d-bus endpoint object monitoring applications about case where MCTP device has fallen off the network. This would greatly help these applications to initiate their initial discovery sequence as early/quickly as possible without losing out valuable telemetry data since they would be made aware of such reset via MCTPD.

If this seems like a legit reason to you for which we can think of keeping polling, I have further question on implementation

The particular use case which might not get solved even with Polling is fast reset of downstream endpoints such as GPUs themselves which happens in time-window of couple of milliseconds (100-200 msec) since polling window is not be narrow enough to capture such fast reset and from MCTP perspective endpoint never gotten off the MCTP network. In such case NSM would not even get notified about this fast-reset of GPU.

Need some help with figuring out how do we expect endpoint to notify or BusOwner to become aware of such fast reset use case.

I'm open to any further guidance/opinion on Polling mechanism itself, or should MCTP even care about this etc. Or may be help me understand this better on the philosophy of mctpd not caring about state of device downstream to bridge or direct-peer itself.

faizana-nvidia commented:
Hello Jeremy,

Updated the PR with last discussion on stopping of polling. It's up for review now.

Sorry to be reluctant on this, but is there any plan to support Discovery Notify for direct peer to BMC as bus owner specifically for something like USB binding where we don't explicitly have methods like I2C bus detection to identify endpoint presence?

jk-ozlabs commented:
Thanks for the updates.

You now seem to have a bunch of fixes/changes to the bridging implementation, which don't belong in the polling feature addition.

Can you separate those into a different PR please?

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1862,6 +1907,13 @@ static int remove_peer(struct peer *peer)
 		sd_event_source_unref(peer->recovery.source);
 	}
 
+	if (peer->pool_size) {
```
ACK, I should remove pctx too here

faizana-nvidia replied:
I'll squash the commits

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4763,6 +4773,139 @@ static int endpoint_send_allocate_endpoint_ids(
+	int i;
+	ctx = bridge->ctx;
+	sd_event_source **sources =
+		calloc(pool_size, sizeof(sd_event_source *));
```
understood, will update

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4763,6 +4773,139 @@ static int endpoint_send_allocate_endpoint_ids(
+
+	if (!bridge) {
+		warnx("Bridge (eid %d) removed?", bridge->eid);
+		goto exit;
```
okay sure

faizana-nvidia commented:
Sorry for  the delay in addressing the review comments. I'll push the changes soon, meanwhile could you help me with my responses

faizana-nvidia reviewed (commented):
I found one problem here, looks like kernel is not deleting the routes by itself for gateway routes. And we would avoid deleting routes if have_route flag is not set.

Should we consider removing from mctpd for now?

> static int del_interface(struct link *link)
{
	struct ctx *ctx = link->ctx;
	int ifindex = link->ifindex;
	if (ctx->verbose) {
		fprintf(stderr, "Deleting interface #%d\n", ifindex);
	}
	for (size_t i = 0; i < ctx->num_peers;) {
		struct peer *p = ctx->peers[i];
		if (p->state == REMOTE && p->phys.ifindex == ifindex) {
			int rc;
			// Linux removes routes to deleted links, so no need
			// to request removal.
			p->have_neigh = false;
			p->have_route = false;

jk-ozlabs commented:
> Is there a way to increase the time frame for unit tests beyond 30sec

Possibly, but I think the better approach here would be to decouple from wall-clock time in the test framework. There has been some work on that recently, but I don't think we're at a point where that's solved.

I'll take a look; if that's going to be longer-term then we may just want to temporarily increase the timeout...

faizana-nvidia commented:
> > Is there a way to increase the time frame for unit tests beyond 30sec
> 
> Possibly, but I think the better approach here would be to decouple from wall-clock time in the test framework. There has been some work on that recently, but I don't think we're at a point where that's solved.
> 
> I'll take a look; if that's going to be longer-term then we may just want to temporarily increase the timeout...

Hello @jk-ozlabs 

Gentle reminder, Will this block us for re-review this PR?

jk-ozlabs commented:
OK, the timers support has been merged. Let me know how you go with that.

We may want to discuss interactions with Get Routing Table Entries too.

faizana-nvidia commented:
> OK, the timers support has been merged. Let me know how you go with that.
> 

It worked like a charm :), test cases are running pretty fast. PR has been updated
> We may want to discuss interactions with Get Routing Table Entries too.

I see the PR [128](https://github.com/CodeConstruct/mctp/pull/128) opened for GRTE command support, will go though it

Pulling in your conversation here https://github.com/CodeConstruct/mctp/pull/128#issuecomment-3471123399

> _Perhaps we could use the GRTE responses as first-pass data on what to start polling, plus a configuration for "just poll the entire downstream pool" for cases where bridges do not support (or do not reliably support) GRTE_

For interacting with GRTE, as you mentioned in your above comment, idea is, to reply on polling only when GRTE isn't supported by Bridge.

Question :

If GRTE Supported - should polling continue only for eids (among the allocated pool space)which are not present as routing entry of bridge's routing table? if so then what happens for the case  where some endpoint (among the pool range)starts responding to polling commands, will we ask bridge to update its routes (may be using Routing Information Update)

If not Supported - continue as this PR is doing.

jk-ozlabs commented:
Yep, I think that's good summary of the logic there :)

Since it's an optimisation change (reducing polling requirements) rather than correctness (we'll still publish the saame enumeration results), I am OK with separating the GRTE changes until after the polling implementation, if you prefer.

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -4752,6 +4756,26 @@ static int parse_config_bus_owner(struct ctx *ctx, toml_table_t *bus_owner)
+		uint64_t i = val.u.i;
+		if (i == 0) {
+			ctx->endpoint_poll = 0;
+			fprintf(stderr, "Bridge poll is disabled\n");
```
Super minor, but we probably don't need this output by default. You could make it conditional on `ctx->verbose`, or remove it altogether.

faizana-nvidia replied:
Ack,  l'll make it conditional

faizana-nvidia commented:
> I think the structure here is sensible, just a few queries and changes inline.

Thanks for the review comments, I'll address to them shortly

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -2355,6 +2355,52 @@ static int get_endpoint_peer(struct ctx *ctx, sd_bus_error *berr,
+ * since bridged endpoints are routed from bridge, direct query
+ * to eid should work if gateway routes are in place.
+ */
+static int query_endpoint_poll_commmand(struct peer *peer, mctp_eid_t *resp_eid)
```
right, let me merge the changes

faizana-nvidia commented on tests/mctpenv/__init__.py:
```diff
@@ -259,11 +259,21 @@ def find_endpoint(self, addr):
+                # for the gated endpoints, but only need to find the
+                # gateway's physical address
+                gw_net, gw_eid = route.gw
+                gw_route = self.lookup_route(gw_net, gw_eid)
```
Although for the above case in test framework, we would need to find a recursive lookup among Endpoint's bridged_eid so pass on incoming packet to expected downstream eid 

This is for forwarding the packets:  first based on routes find the interface and eid this needs to go toand then to right Bridge eid which will further forward to the right endpoint

jk-ozlabs replied:
> Wondering if we would really see this recursive gateway route case.

It's certainly the most common case to have the one gateway route handling the entire pool. However, the kernel route implemenation does allow for nesting, to allow us to explicitly describe the full routing topology. For example, this may be useful to describe cases where a bridge (and its subordinate devices) can be routable over different paths. In that case, we can describe that alternate route - possibly with a higher priority - exactly as the hardware connections work.

We don't set up routes like that in mctpd currently though, so perhaps support for that would be OK to defer. But if we do that, we should indicate with a TODO comment.

jk-ozlabs commented on conf/mctpd.conf:
```diff
@@ -16,3 +16,6 @@ message_timeout_ms = 30
 max_pool_size = 15
+
+# Bus Owner/bridge polling interval (ms) to check endpoint accessibility.
+endpoint_poll_ms = 0 # disable poll
```
Super minor, but maybe document the 0 = disable logic in the top-level comment, so that it isn't lost when someone actually configures the value. Perhaps:

```
# Bus Owner/bridge polling interval (in milliseconds) to check endpoint acccessibility
# A value of 0 will disable polling
endpoint_poll_ms = 0
```

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -5213,6 +5328,139 @@ static int endpoint_send_allocate_endpoint_ids(
+	}
+
+	rc = query_endpoint_poll_commmand(peer, &ret_eid);
+	if (rc < 0) {
```
Should we reschedule on all failures, or just timeouts?

faizana-nvidia replied:
gentle reminder for this query @jk-ozlabs

jk-ozlabs replied:
Isn't that the opposite of what I had suggested?

I would be okay with that, with sufficient reasoning. Why should we not reschedule on other errors?

faizana-nvidia replied:
> Should we reschedule on all failures, or just timeouts?

Sorry I got the impression we don't want reschedule for all errors (which I had in my previous version).
Reason why I felt this being okay was, like you stated if for unsuccessful case (for whatever reason) get endpoint id fails, likely next consecutive commands will also fail.

Im okay to update it to reschedule for other error cases as well owing to possibility of recovering in cases, such as :
some mismatch response which might have come in between 
or some send failure in the socket etc

faizana-nvidia replied:
updated to reschedule for all errors

faizana-nvidia replied:
yes, that's kept intentional, lemme know if you see any concern there.

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -1963,6 +1974,57 @@ static int check_peer_struct(const struct peer *peer, const struct net *n)
+		int idx = ep - pool_start;
+
+		if (bridge->bridge_ep_poll.sources &&
+		    bridge->bridge_ep_poll.sources[idx]) {
```
You have many uses of `bridge->bridge_ep_poll.sources[idx]` here, use a temporary.

... and no need to repeat the check for `bridge->bridge_ep_poll.sources` on every iteration, just do it once.

faizana-nvidia replied:
Ack

jk-ozlabs commented on CHANGELOG.md:
```diff
@@ -59,6 +59,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 9. `mctpd`'s Network1.LearnEndpoint method will now check for endpoint
    presence before publishing an endpoint.
 
+10. `mctpd` now implements bridge endpoint polling after successful dynamic pool
```
This appears to have been added to the previous release, we want it in the top section.

For that reason: If you prefer, I am happy to manage the CHANGELOG entries. But if you leave this commit here, there needs to be a space after the `:` in the commit title.

faizana-nvidia replied:
Thats fine, I'll just remove this commit, I feel you could manage the release info better.

jk-ozlabs reviewed (changes requested):
Mostly minor things there, but we do need to clarify the addition of an peer (through `add_peer`) before we know that the peer actually exists. We may not need the struct peer during polling at all.

Also, a couple of typos in the commit log: `endpoin` and `repond`. I don't have a facility to comment on the changelogs themselves, sorry!

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1963,6 +1974,57 @@ static int check_peer_struct(const struct peer *peer, const struct net *n)
+/* Stops downstream endpoint polling and removes
+ * peer structure when bridge endpoint is being removed.
+ */
+static int remove_bridged_peers(struct peer *bridge)
```
its generic method to purge out bridged endpoint when bridge is going away, I should rather remove comments to avoid the confusion

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -5213,6 +5328,139 @@ static int endpoint_send_allocate_endpoint_ids(
+	n = lookup_net(bridge->ctx, bridge->net);
+	peer = n->peers[ep];
+	if (!peer) {
+		rc = add_peer(bridge->ctx, &(bridge->phys), ep, bridge->net,
```
> we don't have much option then, need to utilise raw method _endpoint_query_addr()

Yes, I think that may be best.

however, not sure I understand:

> may be create a new poll method which does entire mctp structure population then sends it over,

What "entire mctp structure" is this referring to? We can use the poll method that you have defined here (`peer_endpoint_poll`), but using `endpoint_query_addr` as the actual transfer mechanism. No?

jk-ozlabs commented:
Merged, thanks!

faizana-nvidia commented:
> Merged, thanks!

Thank you for all the help with code reviews and design discussions  :)
